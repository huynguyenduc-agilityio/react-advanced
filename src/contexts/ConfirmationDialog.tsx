import { createContext, ReactNode, useContext, useState } from 'react';
import { useDisclosure } from '@chakra-ui/react';

type DialogProps = {
  title: string;
  confirmMessage: ReactNode;
  onConfirm: () => void;
  onCancel?: () => void;
};

const initialValue: DialogProps = {
  title: '',
  confirmMessage: '',
  onCancel: undefined,
  onConfirm: () => {},
};

type ContextDefaultValue = {
  confirmInfo: DialogProps;
  confirm: (confirmInfo: DialogProps) => void;
  isOpen: boolean;
  onClose: () => void;
};

const ctxDefaultValue: ContextDefaultValue = {
  confirmInfo: initialValue,
  confirm: () => {},
  isOpen: false,
  onClose: () => {},
};

const ConfirmationDialogContext = createContext(ctxDefaultValue);

const useConfirmationDialog = () => useContext(ConfirmationDialogContext);

const ConfirmationDialogProvider = ({ children }: { children: ReactNode }) => {
  const [confirmInfo, setConfirmInfo] = useState(initialValue);
  const { isOpen, onOpen, onClose } = useDisclosure();

  const handleConfirm = (dialogInfo: DialogProps) => {
    setConfirmInfo(dialogInfo);
    onOpen();
  };

  return (
    <ConfirmationDialogContext.Provider
      value={{ confirmInfo, confirm: handleConfirm, isOpen, onClose }}
    >
      {children}
    </ConfirmationDialogContext.Provider>
  );
};

export { ConfirmationDialogProvider, useConfirmationDialog };
