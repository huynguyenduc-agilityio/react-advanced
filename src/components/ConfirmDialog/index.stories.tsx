import type { Meta, StoryObj } from '@storybook/react';
import { Button } from '@chakra-ui/react';

// Contexts
import { ConfirmationDialogProvider, useConfirmationDialog } from '@/contexts';

import ConfirmDialog from '.';

const meta: Meta<typeof ConfirmDialog> = {
  title: 'Components/ConfirmDialog',
  component: ConfirmDialog,
  parameters: {
    layout: 'centered',
  },

  tags: ['autodocs'],
  decorators: [
    (Story) => (
      <ConfirmationDialogProvider>
        <Story />
      </ConfirmationDialogProvider>
    ),
  ],
};

export default meta;

type Story = StoryObj<typeof ConfirmDialog>;

const Template = () => {
  const { confirm } = useConfirmationDialog();

  const handleConfirmDialog = () =>
    confirm({
      title: 'Delete Item',
      confirmMessage: 'Are you sure you want to delete this item?',
      onConfirm: () => alert('Confirmed!'),
      onCancel: () => alert('Cancelled!'),
    });

  return (
    <>
      <Button onClick={handleConfirmDialog}>Delete Item</Button>
      <ConfirmDialog />
    </>
  );
};

export const Default: Story = {
  render: Template,
};
