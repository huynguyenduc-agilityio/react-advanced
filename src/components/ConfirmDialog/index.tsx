import React from 'react';
import {
  AlertDialog,
  AlertDialogBody,
  AlertDialogContent,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogOverlay,
  Button,
} from '@chakra-ui/react';

// Contexts
import { useConfirmationDialog } from '@/contexts';

const ConfirmDialog = () => {
  const { isOpen, onClose, confirmInfo } = useConfirmationDialog();
  const cancelRef = React.useRef<HTMLButtonElement>(null);

  const handleConfirm = () => {
    confirmInfo.onConfirm();
    onClose();
  };

  const handleCancel = () => {
    confirmInfo.onCancel?.();
    onClose();
  };

  return (
    <AlertDialog
      motionPreset="slideInBottom"
      isCentered
      isOpen={isOpen}
      leastDestructiveRef={cancelRef}
      onClose={onClose}
    >
      <AlertDialogOverlay>
        <AlertDialogContent bgColor="midNightBlue">
          <AlertDialogHeader fontSize="lg" fontWeight="bold" color="white">
            {confirmInfo.title}
          </AlertDialogHeader>

          <AlertDialogBody
            color="pastelBlue"
            fontSize="md"
            px="35px"
            py="24px"
            borderY="0.6px solid"
            borderColor="slateBlue"
          >
            {confirmInfo.confirmMessage}
          </AlertDialogBody>

          <AlertDialogFooter>
            <Button
              variant="outline"
              fontSize="md"
              ref={cancelRef}
              onClick={handleCancel}
              fontWeight="bold"
            >
              Cancel
            </Button>
            <Button
              variant="primary"
              fontSize="md"
              onClick={handleConfirm}
              ml={3}
              fontWeight="bold"
            >
              Confirm
            </Button>
          </AlertDialogFooter>
        </AlertDialogContent>
      </AlertDialogOverlay>
    </AlertDialog>
  );
};

export default ConfirmDialog;
