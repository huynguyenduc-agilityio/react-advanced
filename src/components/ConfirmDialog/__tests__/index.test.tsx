import { ChakraProvider } from '@chakra-ui/react';
import { fireEvent, render, screen } from '@testing-library/react';

import { useConfirmationDialog } from '@/contexts';

import ConfirmDialog from '..';

// Mock the `useConfirmationDialog` context
jest.mock('@/contexts', () => ({
  useConfirmationDialog: jest.fn(),
}));

describe('ConfirmDialog', () => {
  const mockOnConfirm = jest.fn();
  const mockOnCancel = jest.fn();

  beforeEach(() => {
    // Mock the context for each test
    (useConfirmationDialog as jest.Mock).mockReturnValue({
      isOpen: true,
      onClose: jest.fn(),
      confirmInfo: {
        title: 'Confirm Action',
        confirmMessage: 'Are you sure you want to proceed?',
        onConfirm: mockOnConfirm,
        onCancel: mockOnCancel,
      },
    });
  });

  it('renders the dialog with title and message', () => {
    render(
      <ChakraProvider>
        <ConfirmDialog />
      </ChakraProvider>,
    );

    // Check if title and message are rendered
    expect(screen.getByText('Confirm Action')).toBeInTheDocument();
    expect(
      screen.getByText('Are you sure you want to proceed?'),
    ).toBeInTheDocument();
  });

  it('calls onConfirm when Confirm button is clicked', () => {
    render(
      <ChakraProvider>
        <ConfirmDialog />
      </ChakraProvider>,
    );

    fireEvent.click(screen.getByText('Confirm'));
    expect(mockOnConfirm).toHaveBeenCalled();
  });

  it('calls onCancel when Cancel button is clicked', () => {
    render(
      <ChakraProvider>
        <ConfirmDialog />
      </ChakraProvider>,
    );

    fireEvent.click(screen.getByText('Cancel'));
    expect(mockOnCancel).toHaveBeenCalled();
  });

  it('should match snapshot', () => {
    const { asFragment } = render(
      <ChakraProvider>
        <ConfirmDialog />
      </ChakraProvider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
