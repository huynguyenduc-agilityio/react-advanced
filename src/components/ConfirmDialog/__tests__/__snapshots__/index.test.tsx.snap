// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmDialog should match snapshot 1`] = `
<DocumentFragment>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
